src/rust/engine/rule_graph/README.md

```diff
@@ -0,0 +1,119 @@
+
+But both of the goals are important, because together they allow for an API that is* (should be: see the "Issue Overview" section) very easy to write `Rule`s for, with minimal boilerplate required to get the inputs needed for a `Rule` to compute a value, and minimal invalidation. Because the identity of a `Rule` is computed from its transitive input `Param`s rather than from its positional arguments, `Rule`s can accept arbitrarily-many large input values (which don't need to implement hash) with no impact on its memoization hit rate.
+
+On the other hand, we _are_ willing to add constraints in order to make the graph easier to compute. For example, one constraint (which aligns reasonably well with user's expectations) is that only `Rule` dependencies that use the input `Param` of a `Get` are eligible to be used: so a `Get(X, Y)` (`Get` an output type of `X` for an input type of `Y`) will only be satisfied by a subgraph that actually consumes `Y`.
+
+## Implementation
+
+As of [3a188a1e06](https://github.com/pantsbuild/pants/blob/3a188a1e06d8c27ff86d8c311ff1b2bdea0d39ff/src/rust/engine/rule_graph/src/builder.rs#L202-L219), we construct a `RuleGraph` using a combination of data flow analysis and some homegrown (and likely problematic: see the "Issue Overview") node splitting on the call graph of `Rule`s.
+
+The construction algorithm is broken up into phases:
+
+1. [initial_polymorphic](https://github.com/pantsbuild/pants/blob/3a188a1e06d8c27ff86d8c311ff1b2bdea0d39ff/src/rust/engine/rule_graph/src/builder.rs#L221) - Building a polymorphic graph, where all `Rule`s or `Param`s that compute a particular output type are included, but without regard for which `Param`s are available at the use site. This phase will fail fast if no `Query` or `Get` anywhere in the graph provides a particular `Param` type. During this phase nodes may have multiple dependency edges per `DependencyKey`, which is what makes them "polymorphic": each of the possible ways to compute a dependency will likely have different input `Param` requirements, and each node in this phase represents all of them.
+2. [live_param_labeled](https://github.com/pantsbuild/pants/blob/3a188a1e06d8c27ff86d8c311ff1b2bdea0d39ff/src/rust/engine/rule_graph/src/builder.rs#L749-L754) - Run [live variable analysis](https://en.wikipedia.org/wiki/Live_variable_analysis) on the polymorphic graph to compute the initial set of `Params` used by each node in the graph. Because nodes in the polymorphic graph node have references to all possible sources of a particular dependency type, the computed set is conservative (i.e., overly large).
```

Comment:
```suggestion
2. [live_param_labeled](https://github.com/pantsbuild/pants/blob/3a188a1e06d8c27ff86d8c311ff1b2bdea0d39ff/src/rust/engine/rule_graph/src/builder.rs#L749-L754) - Run [live variable analysis](https://en.wikipedia.org/wiki/Live_variable_analysis) on the polymorphic graph to compute the initial set of `Params` used by each node in the graph. Because nodes in the polymorphic graph have references to all possible sources of a particular dependency type, the computed set is conservative (i.e., overly large).
```